esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+				w.flush()
+			}
+			w.mu.Unlock()
+		}
+	}()
+}
+
+// writeMeta formats and writes the item metadata to the buffer; it must be called under a lock.
+//
+func (w *worker) writeMeta(item BulkIndexerItem) error {
+	// TODO(karmi): Handle errors
+	w.buf.WriteRune('{')
+	w.aux = strconv.AppendQuote(w.aux, item.Action)
+	w.buf.Write(w.aux)
+	w.aux = w.aux[:0]
+	w.buf.WriteRune(':')
+	w.buf.WriteRune('{')
+	if item.DocumentID != "" {
+		w.buf.WriteString(`"_id":`)
+		w.aux = strconv.AppendQuote(w.aux, item.DocumentID)
+		w.buf.Write(w.aux)
+		w.aux = w.aux[:0]
+	}
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('\n')
+	return nil
+}
+
+// writeBody writes the item body to the buffer; it must be called under a lock.
+//
+func (w *worker) writeBody(item BulkIndexerItem) error {
+	// TODO(karmi): Handle errors
+	if item.Body != nil {
+		if _, err := w.buf.ReadFrom(item.Body); err != nil {
+			return err
+		}
+		w.buf.WriteRune('\n')
+	}
+	return nil
+}
+
+// flush writes out the worker buffer; it must be called under a lock.
+//
+func (w *worker) flush() error {
+	if w.isFlushing {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: Already flushing\n", w.id)
+		}
+		return nil
+	}
+
+	if w.buf.Len() < 1 {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: Buffer empty\n", w.id)
+		}
+		return nil
+	}
+
+	var (
+		err error
+		blk BulkIndexerResponse
+		ctx = context.Background()
+	)
+
+	defer func() {
+		w.items = w.items[:0]
+		w.buf.Reset()
+		w.isFlushing = false
+	}()
+
+	w.isFlushing = true
+
+	if w.bi.config.DebugLogger != nil {
+		w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: %s\n", w.id, w.buf.String())
+	}
+
+	atomic.AddUint64(&w.bi.stats.numRequests, 1)
+	req := esapi.BulkRequest{
+		Index: w.bi.config.Index,
+		Body:  w.buf,
+
+		Pipeline:            w.bi.config.Pipeline,
+		Refresh:             w.bi.config.Refresh,
+		Routing:             w.bi.config.Routing,
+		Source:              w.bi.config.Source,
+		SourceExcludes:      w.bi.config.SourceExcludes,
+		SourceIncludes:      w.bi.config.SourceIncludes,
+		Timeout:             w.bi.config.Timeout,
+		WaitForActiveShards: w.bi.config.WaitForActiveShards,
+
+		Pretty:     w.bi.config.Pretty,
+		Human:      w.bi.config.Human,
+		ErrorTrace: w.bi.config.ErrorTrace,
+		FilterPath: w.bi.config.FilterPath,
+		Header:     w.bi.config.Header,
+	}
+	res, err := req.Do(ctx, w.bi.config.Client)
+	if err != nil {
+		atomic.AddUint64(&w.bi.stats.numFailed, uint64(len(w.items)))
+		// TODO(karmi): Wrap error
+		return fmt.Errorf("flush: %s", err)
+	}
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	if res.IsError() {
+		atomic.AddUint64(&w.bi.stats.numFailed, uint64(len(w.items)))
+		// TODO(karmi): Wrap error (include response struct)
+		return fmt.Errorf("flush: %s", res.String())
+	}
+
+	if err := w.bi.config.Decoder.UnmarshalFromReader(res.Body, &blk); err != nil {
```

Comment:
I see a lot of value in exposing a configurable JSON decoder; let me explain the various points:

* As of now, the package is [completely dependency-free](https://github.com/elastic/go-elasticsearch/blob/master/go.mod), with only the stdlib packages being imported. I would try my best to keep it that way. (Notice the demonstrations in `_examples` use a lot of third-party packages, but they are not part of the package itself.) Of course, this is not a dogma, but I haven't yet found a good-enough reason to break this commitment. In other words, it doesn't feel right to me to add a third-party dependency to the entire package for something such as parsing an HTTP response in the `esutil` sub-package.

* Furthermore, I see a configurable decoder as a fundamental feature of the package as a whole. For context, it did serve us very good for the original client architectures in Ruby and Python — a configurable `Serializer` component is something a lot of people find use for. (See eg. [this example](https://gist.github.com/lribeiro/7d9fbedf830a54685811f63dbbce9464) for implementing a custom serializer for Python's `numpy` library).

* I expect to use a strategy like this for any other utility component which has to deal with JSON encoding/decoding, like a possible `Search()` or `Scan()`  helper.

Reply:
Right, I see. I'm not against what you're suggesting here, but I guess I'm just not sure how to implement the JSON parsing for this well-defined structure without explicitely — or implicitely — relying on a third-party package. Do you have a suggestion for that?

In general, JSON parsing by the standard library _is_ quite a severe bottleneck — I would have to run the benchmarks again to be able to provide some numbers, but I did see a drop in performance there.

Is your concern mainly about exposing a part of the public API — the `Decoder` component —, unneccessarily?

Reply:
> Right, I see. I'm not against what you're suggesting here, but I guess I'm just not sure how to implement the JSON parsing for this well-defined structure without explicitely — or implicitely — relying on a third-party package. Do you have a suggestion for that?

Not yet. It'll probably involve reinventing the wheel - i.e. yet another decoder generator, but one that generates all the required code. How feasible that is I'm not yet sure.

> Is your concern mainly about exposing a part of the public API — the Decoder component —, unneccessarily?

Yes.

Reply:
I see, thanks for the clarification.

I don't think that we can reinvent the wheel in a matter of days :) Because I wouldn't like to ship this feature with a known performance degradation, coming from `encoding/json`, I would rather the `Decoder` extension point as is, and deprecate, or even straight out remove it later.